labels.go

```diff
@@ -0,0 +1,299 @@
+// that's more restrictive than absolutely necessary.
+//
+// Tally allows anything matching the regexp `^[0-9A-z_\-]+$`. Prometheus
+// allows the regexp `^[A-z_:][0-9A-z_:]*$` for metric names, and
+// `^[A-z_][0-9A-z_]*$` for label names.
+//
+// The common subset is `^[A-z_][0-9A-z_]*$`.
+func IsValidName(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+
+	switch c := s[0]; {
+	case 'A' <= c && c <= 'Z':
+		break
+	case 'a' <= c && c <= 'z':
+		break
+	case c == '_':
+		break
+	default:
+		return false
+	}
+
+	// Don't incur the expense of ranging over runes, since no multibyte UTF-8
+	// characters are legal.
+	for i := 1; i < len(s); i++ {
+		c := s[i]
+		switch {
+		case '0' <= c && c <= '9':
+			continue
+		case 'A' <= c && c <= 'Z':
+			continue
+		case 'a' <= c && c <= 'z':
+			continue
+		case c == '_':
+			continue
+		default:
+			return false
+		}
+	}
+
+	return true
+}
+
+// IsValidLabelValue checks whether the supplied string is a valid label value
+// in both Prometheus and Tally.
+//
+// Tally allows label values that match the regexp `^[0-9A-z_\-.]+$`.
+// Prometheus allows any valid UTF-8 string.
+func IsValidLabelValue(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+
+	// Don't incur the expense of ranging over runes, since no multibyte UTF-8
+	// characters are legal.
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		switch {
+		case '0' <= c && c <= '9':
+			continue
+		case 'A' <= c && c <= 'Z':
+			continue
+		case 'a' <= c && c <= 'z':
+			continue
+		case c == '_' || c == '-' || c == '.':
+			continue
+		default:
+			return false
+		}
+	}
+	return true
+}
+
+// scrubLabelValue replaces any invalid runes in the input string with '_'. If
+// the input is already a valid label value (in both Prometheus and Tally),
+// it's returned unchanged.
+func scrubLabelValue(s string) string {
+	if IsValidLabelValue(s) {
+		return s
+	}
+	if len(s) == 0 {
+		return DefaultLabelValue
+	}
+
+	d := newDigester()
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		switch {
+		case '0' <= c && c <= '9':
+			d.bs = append(d.bs, c)
+		case 'A' <= c && c <= 'Z':
+			d.bs = append(d.bs, c)
+		case 'a' <= c && c <= 'z':
+			d.bs = append(d.bs, c)
+		case c == '_' || c == '-' || c == '.':
+			d.bs = append(d.bs, c)
+		default:
+			d.bs = append(d.bs, '_')
+		}
+	}
+	scrubbed := string(d.bs)
+	d.free()
+	return scrubbed
+}
+
+func areValidLabelValues(ss []string) bool {
+	for _, s := range ss {
+		if !IsValidLabelValue(s) {
+			return false
+		}
+	}
+	return true
+}
+
+func scrubLabelValues(ss []string) []string {
+	if areValidLabelValues(ss) {
+		return ss
+	}
+
+	scrubbed := make([]string, len(ss))
+	for i, dirty := range ss {
+		scrubbed[i] = scrubLabelValue(dirty)
+	}
+	return scrubbed
+}
+
+// scrubName replaces any invalid runes in the input string with '_'. If the
+// input is already a valid label and metric name (in both Prometheus and
+// Tally), it's returned unchanged.
+func scrubName(s string) string {
```

Comment:
suggest moving `scrubName` closer to `IsValidName`

Reply:
Sure thing.